client/client.go

```diff
@@ -0,0 +1,389 @@
+			if err := c.client.Get(ctx, key, u); err != nil {
+				return err
+			}
+			u.SetGroupVersionKind(topolvmv1.GroupVersion.WithKind(kind))
+			return c.scheme.Convert(u, obj, nil)
+		}
+		return c.client.Get(ctx, key, obj)
+	}
+	return c.client.Get(ctx, key, obj)
+}
+
+func (c *wrappedReader) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	gvk := list.GetObjectKind().GroupVersionKind()
+	switch o := list.(type) {
+	case *unstructured.UnstructuredList:
+		if gvk.Group == group && gvk.Kind == kindList && topolvm.UseLegacy() {
+			o.SetGroupVersionKind(topolvmlegacyv1.GroupVersion.WithKind(kindList))
+			err := c.client.List(ctx, list, opts...)
+			o.SetGroupVersionKind(topolvmv1.GroupVersion.WithKind(kindList))
+			return err
+		}
+		return c.client.List(ctx, list, opts...)
+	case *metav1.PartialObjectMetadataList:
+		if gvk.Group == group && gvk.Kind == kindList && topolvm.UseLegacy() {
+			o.SetGroupVersionKind(topolvmlegacyv1.GroupVersion.WithKind(kindList))
+			err := c.client.List(ctx, list, opts...)
+			o.SetGroupVersionKind(topolvmv1.GroupVersion.WithKind(kindList))
+			return err
+		}
+		return c.client.List(ctx, list, opts...)
+	case *topolvmv1.LogicalVolumeList:
+		if topolvm.UseLegacy() {
+			u := &unstructured.UnstructuredList{}
+			if err := c.scheme.Convert(list, u, nil); err != nil {
+				return err
+			}
+			u.SetGroupVersionKind(topolvmlegacyv1.GroupVersion.WithKind(kindList))
+			if err := c.client.List(ctx, u, opts...); err != nil {
+				return err
+			}
+			u.SetGroupVersionKind(topolvmv1.GroupVersion.WithKind(kindList))
+			return c.scheme.Convert(u, list, nil)
+		}
+		return c.client.List(ctx, list, opts...)
+	}
+	return c.client.List(ctx, list, opts...)
+}
+
+type wrappedClient struct {
+	reader client.Reader
+	client client.Client
+}
+
+var _ client.Client = &wrappedClient{}
+
+func NewWrappedClient(c client.Client) client.Client {
+	return &wrappedClient{
+		reader: NewWrappedReader(c, c.Scheme()),
+		client: c,
+	}
+}
+
+func (c *wrappedClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	return c.reader.Get(ctx, key, obj)
+}
+
+func (c *wrappedClient) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	return c.reader.List(ctx, list, opts...)
+}
+
+func (c *wrappedClient) Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	switch o := obj.(type) {
+	case *unstructured.Unstructured:
+		if gvk.Group == group && gvk.Kind == kind && topolvm.UseLegacy() {
+			o.SetGroupVersionKind(topolvmlegacyv1.GroupVersion.WithKind(kind))
+			err := c.client.Create(ctx, o, opts...)
+			o.SetGroupVersionKind(topolvmv1.GroupVersion.WithKind(kind))
+			return err
+		}
+		return c.client.Create(ctx, obj, opts...)
+	case *metav1.PartialObjectMetadata:
+		return c.client.Create(ctx, obj, opts...)
```

Comment:
Sorry for my lack of understanding, but why is the group-and-version check not required for `Create` and `Update`?

Reply:
- `client. Create`
- `client. Update`
- `statusWriter. Update`

because these 3 methods don't support `*metav1.PartialObjectMetadata`.

- https://github.com/kubernetes-sigs/controller-runtime/blob/v0.12.3/pkg/client/client.go#L181-L187
- https://github.com/kubernetes-sigs/controller-runtime/blob/v0.12.3/pkg/client/client.go#L193-L200
- https://github.com/kubernetes-sigs/controller-runtime/blob/v0.12.3/pkg/client/client.go#L304-L310

Reply:
Thanks. I got it.